### PR TITLE
implement glyphsLib's logic to find base style shared by all masters

### PR DESCRIPTION
**Pull request by anthrotype**
Fixes https://github.com/googlefonts/fontc/issues/264

After this [HankenGrotesk-Italic.glyphs](https://github.com/marcologous/hanken-grotesk/blob/master/sources/HankenGrotesk-Italic.glyphs) is happier

### CODE REVIEW

File: glyphs-reader/src/font.rs [hunk 1]
```diff
@@ -2683,6 +2725,46 @@ mod tests {
...
+        &[
+            "Condensed Thin",
+            "Condensed Light",
+            "Condensed Regular",
+        ],
+        2
+    )]
+    #[case::default_to_regular(&["Thin", "Light", "Regular", "Medium", "Bold"], 2)]
+    #[case::default_to_first(&["Foo", "Bar", "Baz"], 0)]
+    fn find_default_master(#[case] master_names: &[&str], #[case] expected_idx: usize) {
```
Comment (rsheeter) [inline glyphs-reader/src/font.rs:2755]: Optional: the cases would be easier to read if the expected answer was also a string

File: glyphs-reader/src/font.rs [hunk 2]
```diff
@@ -1502,17 +1502,54 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
...
+                        .join(" ")
+                })
+                .as_ref()
+                == Some(&base_style)
         })
-        .unwrap_or_default()
+        .unwrap_or_default() // just pick the first
+}
+
+// Direct port of glyphsLib's find_base_style:
```
Comment (rsheeter) [inline glyphs-reader/src/font.rs:1537]: While true this doesn't actually help me understand what the relationship between master names and styles is and that makes interpreting the surrounding code difficult.

Comment (rsheeter) [reply glyphs-reader/src/font.rs:1537]: I think we're saying the base style is the word(s) (defined as stuff delimited by whitespace) that are common in all master names is the base style, maybe we could comment as much and add an example.

File: glyphs-reader/src/font.rs [hunk 3]
```diff
@@ -1502,17 +1502,54 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
...
-        // Still nothing? - just look for one called Regular
-        .or_else(|| {
-            raw_font
-                .font_master
-                .iter()
-                .position(|m| matches!(m.name.as_deref(), Some("Regular")))
+    {
+        return master_idx;
+    }
+    // Look for a master whose style is shared by all masters, defaulting to 'Regular'
```
Comment (rsheeter) [inline glyphs-reader/src/font.rs:1509]: As below, this is factually correct but doesn't really help understand the rationale.

My kneejerk reaction is the code feels more complicated than the underlying operation ... let me think about that a bit. Probably more helpful if it comes with an actual suggestion :)

File: glyphs-reader/src/font.rs (hunk 1 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:2755]: I would also suggest a common on each for why, e.g. 

```rust
    #[case::base_style_without_regular(
        &[
            "Expanded Thin Italic",
            "Expanded Italic",
            "Expanded Bold Italic",
        ],
        1  // common words "Expanded Italic", exact match at [1]
    )]
    #[case::base_style_contains_regular(
        &[
            "Regular Foo Bar",
            "Regular Foo Baz",
            "Regular Foo",
        ],
        2. // common words "Regular Foo", exact match at [2]
    )]
    #[case::base_style_with_regular_omitted(
        &[
            "Condensed Thin",
            "Condensed Light",
            "Condensed Regular",
        ],
        2 // common words "Condensed", matches [2] when "Regular" dropped
    )]
    // No common words, "Regular" wins
    #[case::default_to_regular(&["Thin", "Light", "Regular", "Medium", "Bold"], 2)]
   // No common words, no "Regular", first entry wins
    #[case::default_to_first(&["Foo", "Bar", "Baz"], 0)]
```

File: glyphs-reader/src/font.rs [hunk 4]
```diff
@@ -1492,7 +1492,7 @@ fn parse_codepoint_str(s: &str, radix: u32) -> BTreeSet<u32> {
 fn default_master_idx(raw_font: &RawFont) -> usize {
     // Prefer an explicit origin
     // https://github.com/googlefonts/fontmake-rs/issues/44
-    raw_font
+    if let Some(Some(master_idx)) = raw_font
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:1495]: instead of `Some(Some(` change the second `map` to `and_then`.

File: glyphs-reader/src/font.rs [hunk 5]
```diff
@@ -1502,17 +1502,54 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
                 .enumerate()
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:1502]: not your code, but instead of enumerate + find, we can use the `position` method..

File: glyphs-reader/src/font.rs (hunk 2 above)
Comment (cmyr) [reply glyphs-reader/src/font.rs:1537]: I did also find this pretty confusing, and the logic is also kind of funny but I guess depends highly on the use of standard naming conventions? Agree that a few examples would be useful...

File: glyphs-reader/src/font.rs [hunk 6]
```diff
@@ -1502,17 +1502,54 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
...
+    let master_names: Vec<_> = masters.iter().filter_map(|m| m.name.as_deref()).collect();
+    if master_names.is_empty() {
+        return None;
+    }
+
+    let mut base_style: Vec<_> = master_names[0].split_whitespace().collect();
+
+    for master_name in master_names {
+        let style: Vec<_> = master_name.split_whitespace().collect();
+        base_style.retain(|word| style.contains(word));
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:1549]: no need to allocate here, we can just scan the iterator directly:
```suggestion
        base_style.retain(|word| master_name.whitespace().any(|style| style == word));
```

File: glyphs-reader/src/font.rs [hunk 7]
```diff
@@ -1502,17 +1502,54 @@ fn default_master_idx(raw_font: &RawFont) -> usize {
...
+    }
+
+    Some(base_style.join(" ")).filter(|s| !s.is_empty())
```
Comment (cmyr) [inline glyphs-reader/src/font.rs:1552]: filter should be unnecessary, `split_whitespace` doesn't return empty strings.

(also worth noting[1] just to be academic that this function is lossy if the input contained whitespace that weren't a space)

1. okay it's extremely debatable that this is 'worth noting'

**Review by cmyr — COMMENTED** [review]
various little rust nits inside 🎁

File: glyphs-reader/src/font.rs (hunk 3 above)
Comment (rsheeter) [reply glyphs-reader/src/font.rs:1509]: https://github.com/googlefonts/fontc/pull/1001 attempts to:

1. Implement in terms of picking the best master
1. Expand comments based on parts I found hard to follow
1. Operate on `Vec<&str>` rather than forming String

**Review by rsheeter — APPROVED** [review]
LGTM with a couple of suggestions

File: glyphs-reader/src/font.rs (hunk 1 above)
Comment (anthrotype) [reply glyphs-reader/src/font.rs:2765]: > the cases would be easier to read if the expected answer was also a string

done in 4dcde7ac

File: glyphs-reader/src/font.rs (hunk 4 above)
Comment (anthrotype) [reply glyphs-reader/src/font.rs]: done, thanks!  0834b8b9

File: glyphs-reader/src/font.rs (hunk 5 above)
Comment (anthrotype) [reply glyphs-reader/src/font.rs]: of course, sorry! :) 0b352556

File: glyphs-reader/src/font.rs (hunk 7 above)
Comment (anthrotype) [reply glyphs-reader/src/font.rs]: thanks for the comment. it should no longer apply though, given that #1001 isn't doing that any more

File: glyphs-reader/src/font.rs (hunk 6 above)
Comment (anthrotype) [reply glyphs-reader/src/font.rs]: good to know! since we are now scanning more than once after #1001 I think we do need to allocate
